### PR TITLE
[CORE] Use thread local to hold original plan in ColumnarOverrideRules

### PR DESCRIPTION
**Pull request by ulysses-you**
## What changes were proposed in this pull request?

`ColumnarOverrideRules` will be instanced per Spark session rather than per query, so better to use thread local to hold variable to make it thread safe. Including `oringalPlan` and `isAdaptiveContext`.

The local property of SparkContext is a little bit heavy that it would post it to executor side, so we should cleanup it before execution.

## How was this patch tested?

Pass CI

### CODE REVIEW

**Comment by PHILO-HE** [comment]
@baibaichen, @zzcclp

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 1]
```diff
@@ -732,20 +732,55 @@ case class ColumnarOverrideRules(session: SparkSession)
   @transient private lazy val planChangeLogger = new PlanChangeLogger[SparkPlan]()
 
   // This is an empirical value, may need to be changed for supporting other versions of spark.
-  private val aqeStackTraceIndex = 13
+  private val aqeStackTraceIndex = 14
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:735]: why change to 14 ?

**Comment by zzcclp** [comment]
@zheniantoushipashi please take a review, thanks.

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala (hunk 1 above)
Comment (ulysses-you) [reply gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:735]: this pr adds one more function `setAdaptiveContext` and move `Thread.currentThread.getStackTrace` into it.

**Comment by baibaichen** [comment]
related with https://github.com/oap-project/gluten/pull/3255

**Comment by baibaichen** [comment]
> The local property of SparkContext is a little bit heavy that it would post it to executor side, so just use java thread local.

@ulysses-you 

We need disable or enable gluten per query.  If so, we should add similar mechanism at `ColumnarOverrideRules `, doesn't it messy？

**Comment by ulysses-you** [comment]
@baibaichen @PHILO-HE I changed code based on the discussion

**Review by PHILO-HE — COMMENTED** [review]
Just a trivial comment. Thanks!

**Comment by PHILO-HE** [comment]
@zzcclp, @baibaichen, do you have any comment? Thanks!

**Comment by baibaichen** [comment]
LGTM

**Review by zzcclp — APPROVED** [review]
LGTM

**Pull request merged**